### PR TITLE
[R-package] [docs] add intro vignette (#3946)

### PR DESCRIPTION
**Pull request by jameslamb**
## Description from previous PR

*Description copied from #3946. See https://github.com/microsoft/LightGBM/pull/3946#discussion_r744130067 for an explanation of why #3946's changes were moved to this PR.*

First step towards https://github.com/microsoft/LightGBM/issues/1944.

This initial vignette shows the very basic workflow of LightGBM. As such, it would *not* replace the demo with the corresponding name as that seems rather a guide how to deal with sparse data.

The html looks as in this zip here:

[basic_walkthrough.zip](https://github.com/microsoft/LightGBM/files/5970298/basic_walkthrough.zip)

### Some notes on R vignettes in general

- CRAN does not rebuild the vignette html, but it still runs its R code.
- While it is possible to use a logo, it is non-trivial.
- Would you stick to the same code style as in the package code? I am asking because LightGBM is the only project that use R code with comma first.
- The vignette html will be created automatically when building the package. Since the whole building process of LGB is complex, I have no idea if this works out as I hope... still, our colleagues here https://cran.r-project.org/web/packages/xgboost/index.html were successful.

------------------------

## Notes for Reviewers

See https://github.com/microsoft/LightGBM/pull/3946#pullrequestreview-799415812 for some relevant information needed to review this PR.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
RTD builds are enabled for this branch: https://readthedocs.org/projects/lightgbm/builds/15200979/.

**Comment by jameslamb** [comment]
https://lightgbm.readthedocs.io/en/dev-vignettes/R/reference/ looks ok to me!

image

https://lightgbm.readthedocs.io/en/dev-vignettes/R/articles/index.html

image

https://lightgbm.readthedocs.io/en/dev-vignettes/R/articles/basic_walkthrough.html

image

**Comment by StrikerRUS** [comment]
Yeah, seems vignettes were built correctly, that's great!

However, I have one concern. That build with only one tiny vignette took ~620 sec compared to less than 500 sec for `master` branch. Our hard limit on RTD is 900 sec. I'm afraid that with growing number of vignettes we'll reach that limit quite soon. I wonder is there any workaround to rebuild only vignettes with any changes compared to the previous versions? Or the only way to stay within the limits is to write very simple examples and always cut down training rounds?

**Comment by jameslamb** [comment]
> that build with only one tiny vignette took ~620 sec compared to less than 500 sec for master branch.

I think that the particular vignette added in this PR probably added very little to that timing, and the most of the increase is attributable to the fact that LightGBM has to now be compiled twice (once each on these two lines).

https://github.com/microsoft/LightGBM/blob/aafedd8a1c57432ac3e38e4af236c00c6c54e84c/docs/conf.py#L283-L284

If I'm right about that, then it means that adding additional vignettes won't add a lot of time to those builds. I'll test that theory by pushing a test commit here that introduces a copy of `basic_walkthrough.Rmd`, with a different name.

Other strategies we could use if the amount of time to build vignettes becomes problematic in the future:

* mark all code in vignettes as `eval = FALSE` (https://yihui.org/knitr/options/#code-evaluation) so that now R code in vignettes actually runs (and then change CI processes to force them to run, so they're tested)
* try to limit all vignettes to very small `num_iterations`, `num_leaves`, datasets size, etc.

**Comment by jameslamb** [comment]
I just pushed https://github.com/microsoft/LightGBM/pull/4775/commits/40fb2e2f1982402798776ee44e4ec82fc4644d3d which adds 10 identical copies of `basic_walthrough.Rmd`.

The RTD build took 655s (https://readthedocs.org/projects/lightgbm/builds/15206531/), only about 30s longer than the build with a single vignette (https://readthedocs.org/projects/lightgbm/builds/15203195/).

So I think this is strong evidence of what I mentioned in https://github.com/microsoft/LightGBM/pull/4775#issuecomment-962646760, that most of the added build time in RTD builds from this PR is from re-compiling LightGBM.

As more vignettes are added in #1944 we should consider the strategies for limiting runtime that I mentioned in https://github.com/microsoft/LightGBM/pull/4775#issuecomment-962646760, but I think there's a good amount of room to add new vignettes without risking hitting that RTD 900-second limit.

**Comment by StrikerRUS** [comment]
@jameslamb Thanks for your investigation! I agree we may add some number of new vignettes without any worries.

File: R-package/vignettes/basic_walkthrough.Rmd [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+```{r, echo = FALSE, results = "hide"}
+# Cleanup
+if (file.exists("lightgbm.model")) {
+  file.remove("lightgbm.model")
+}
+```
+
+## References
+
+Ke, Guolin, Qi Meng, Thomas Finley, Taifeng Wang, Wei Chen, Weidong Ma, Qiwei Ye, and Tie-Yan Liu. 2017. "LightGBM: A Highly Efficient Gradient Boosting Decision Tree." In Advances in Neural Information Processing Systems 30 (Nip 2017).
```
Comment (StrikerRUS) [inline R-package/vignettes/basic_walkthrough.Rmd:113]:
```suggestion
Ke, Guolin, Qi Meng, Thomas Finley, Taifeng Wang, Wei Chen, Weidong Ma, Qiwei Ye, and Tie-Yan Liu. 2017. "LightGBM: A Highly Efficient Gradient Boosting Decision Tree." In Advances in Neural Information Processing Systems 30 (NIPS 2017).
```

**Review by StrikerRUS — CHANGES_REQUESTED** [review]
I'm really happy this PR allows to start writing cool vignettes! 🎉 
The only disappointing thing is that tar trick.

Please check some my minor suggestions below:

File: .ci/test_r_package_windows.ps1 [hunk 2]
```diff
@@ -165,7 +165,16 @@ if ($env:COMPILER -ne "MSVC") {
...
     Run-R-Code-Redirect-Stderr "commandArgs <- function(...){$env:BUILD_R_FLAGS}; source('build_r.R')"; Check-Output $?
   } elseif ($env:R_BUILD_TYPE -eq "cran") {
+    # NOTE: gzip and tar are needed to create a CRAN package on Windows, but
+    # some flavors of tar.exe can fail in some settings on Windows.
+    # Putting the msys64 utilities at the beginning of PATH temporarily to be
+    # sure they're used for that purpose.
+    # $env:PATH = "C:\msys64\usr\bin;" + $env:PATH
+    if ($env:R_MAJOR_VERSION -eq "3") {
+        $env:PATH = "C:\msys64\usr\bin;" + $env:PATH
+    }
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply .ci/test_r_package_windows.ps1:175]: ah! didn't realize I'd left that behind. Thank you

File: .vsts-ci.yml [hunk 3]
```diff
@@ -300,6 +300,7 @@ jobs:
   steps:
   - script: |
       LGB_VER=$(head -n 1 VERSION.txt | sed "s/rc/-/g")
+      Rscript -e "install.packages(c('R6', 'data.table', 'jsonlite', 'knitr', 'Matrix', 'rmarkdown', 'testthat'), dependencies = c('Depends', 'Imports', 'LinkingTo'), repos = 'https://cran.r-project.org', Ncpus = parallel::detectCores())" || exit -1
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply .vsts-ci.yml:303]: you're right! This was bad copy-pasting from other jobs.

**Comment by jameslamb** [comment]
> the only disappointing thing is that tar trick

agree 😢 

Following https://www.r-project.org/bugs.html, I've request access to R's issue tracker. If that's approved, I'll open a bug report (and maybe even attempt a patch), based on the information in the third item mentioned at https://github.com/microsoft/LightGBM/pull/3946#pullrequestreview-799415812.

**Review by StrikerRUS — APPROVED** [review]
LGTM!

**Comment by jameslamb** [comment]
Thanks @StrikerRUS !

I just updated to latest `master`, just to do one more round of tests. If those don't show any issues, I'll merge this.

**Comment by jameslamb** [comment]
Thanks for this awesome contribution @mayer79 ! I'm sorry it took so long to get this (#3946) merged.

**Comment by mayer79** [comment]
Awesome @jameslamb ! We can soon start adding more vignettes. Should I open an issue with a list of possible topics?

On CRAN, vignettes are sorted by the file name of the corresponding .rmd file, not by the vignette title or its `%\VignetteIndexEntry{[...]}` in the yaml header.

An example from our `data.table` colleagues:
- Order of vignettes on CRAN: https://cran.r-project.org/web/packages/data.table/index.html
- .rmd file names: https://github.com/Rdatatable/data.table/tree/master/vignettes

**Comment by jameslamb** [comment]
thanks for the great information @mayer79 ! I just responded over at https://github.com/microsoft/LightGBM/issues/1944#issuecomment-986453621. Let's move the conversation there for greater visibility.

**Pull request merged**